pt_miniscreen/app.py

```diff
@@ -94,151 +28,41 @@ def set_is_user_controlled(user_has_control) -> None:
-                logger.info("Button press before State manager initialised")
-                return
-
-            if not self.state_manager.buttons_should_be_handled():
-                logger.info("State manager says that buttons should not be handled")
-                return
-
-            self.state_manager.wake()
-
-            handler = {
-                MiniscreenButton.CANCEL: self.current_tile_group.handle_cancel_btn,
-                MiniscreenButton.SELECT: self.current_tile_group.handle_select_btn,
-                MiniscreenButton.UP: self.current_tile_group.handle_up_btn,
-                MiniscreenButton.DOWN: self.current_tile_group.handle_down_btn,
-            }[button]
-
-            if not handler() and button == MiniscreenButton.CANCEL:
-                logger.debug(
-                    "Button press not handled by current tile group - going to next tile group"
-                )
-                self.pop_tile_group()
-
-        self.miniscreen.cancel_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.CANCEL
-        )
-        self.miniscreen.select_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.SELECT
-        )
-        self.miniscreen.up_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.UP
-        )
-        self.miniscreen.down_button.when_released = lambda: handle_button_press(
-            MiniscreenButton.DOWN
-        )
+    def handle_cancel_button_release(self):
+        self.root.exit_menu()
 
-    @property
-    def current_tile_group(self):
-        return self.tile_group_stack[-1]
-
-    def start(self) -> None:
-        self.__stop_event = Event()
-
-        logger.debug("Configuring interrupt signals...")
-        signal(SIGINT, lambda signal, frame: self.stop())
-        signal(SIGTERM, lambda signal, frame: self.stop())
-
-        logger.debug("Starting main app thread...")
-        self.__thread = Thread(target=self._main, args=(), daemon=True)
-        self.__thread.start()
-
-    def wait_for_stop(self) -> None:
-        self.__stop_event.wait()
-        error = getattr(self, "_stop_error", None)
-        if isinstance(error, Exception):
-            raise error
-
-    def stop(self, error: Optional[Exception] = None) -> None:
-        logger.info("Stopping app...")
-        self._stop_error = error
-        self.__stop_event.set()
-        self.__thread.join()
-        self.state_manager.cleanup()
-        unsubscribe_all()
+    def handle_up_button_release(self):
+        self.root.scroll_up()
+
+    def handle_down_button_release(self):
+        self.root.scroll_down()
+
+    def display(self):
+        if self.user_has_control:
+            logger.info("User has control. Waiting for user to give control back...")
+            self.user_gave_back_control_event.wait()
+            self.user_gave_back_control_event.clear()
+            self.display()
```

Comment:
why call `self.display()` again here? this will lead to eventually calling super.display() repeatedly. if there's a need to check user_has_control again, then we should `return self.display()` to prevent displaying repeatedly. but I don't see why we would need to check again, so I think this line can be removed?

Reply:
It reruns the method after the 'gave_back_control' event is set and self.user_has_control is no longer true. I could have just fallen through though which would be less confusing anyways!

Reply:
It does seem that returning control to this app is broken with this implementation as well.